src/centrality.rs

```diff
@@ -0,0 +1,298 @@
+) {
+    let mut do_scale = true;
+    let mut scale = 1.0;
+    if normalized {
+        if endpoints {
+            if node_count < 2 {
+                do_scale = false;
+            } else {
+                scale = 1.0 / (node_count * (node_count - 1)) as f64;
+            }
+        } else if node_count <= 2 {
+            do_scale = false;
+        } else {
+            scale = 1.0 / ((node_count - 1) * (node_count - 2)) as f64;
+        }
+    } else if !directed {
+        scale = 0.5;
+    } else {
+        do_scale = false;
+    }
+    if do_scale {
+        for x in betweenness.iter_mut() {
+            *x = x.map(|y| y * scale);
+        }
+    }
+}
+
+fn _accumulate_basic(
+    betweenness: &mut Vec<Option<f64>>,
+    max_index: usize,
+    path_calc: &mut ShortestPathData,
+    is: usize,
+) {
+    let mut delta = vec![0.0; max_index];
+    for w in &path_calc.verts_sorted_by_distance {
+        let iw = w.index();
+        let coeff = (1.0 + delta[iw]) / (path_calc.sigma[&iw] as f64);
+        let p_w = path_calc.predecessors.get_mut(w).unwrap();
+        for v in p_w {
+            let iv = (*v).index();
+            delta[iv] += (path_calc.sigma[&iv] as f64) * coeff;
+        }
+        if iw != is {
+            betweenness[iw] = betweenness[iw].map(|x| x + delta[iw]);
+        }
+    }
+}
+
+fn _accumulate_endpoints(
+    betweenness: &mut Vec<Option<f64>>,
+    max_index: usize,
+    path_calc: &mut ShortestPathData,
+    is: usize,
+) {
+    if betweenness[is].is_some() {
+        betweenness[is] = Some(
+            betweenness[is].unwrap()
+                + ((path_calc.verts_sorted_by_distance.len() - 1) as f64),
+        );
+    }
+    let mut delta = vec![0.0; max_index];
+    for w in &path_calc.verts_sorted_by_distance {
+        let iw = w.index();
+        let coeff = (1.0 + delta[iw]) / (path_calc.sigma[&iw] as f64);
+        let p_w = path_calc.predecessors.get_mut(w).unwrap();
+        for v in p_w {
+            let iv = (*v).index();
+            delta[iv] += (path_calc.sigma[&iv] as f64) * coeff;
+        }
+        if iw != is {
+            betweenness[iw] = betweenness[iw].map(|x| x + delta[iw] + 1.0);
+        }
+    }
+}
+
+struct ShortestPathData {
+    verts_sorted_by_distance: Vec<NodeIndex>,
+    predecessors: HashMap<NodeIndex, Vec<NodeIndex>>,
+    sigma: HashMap<usize, i64>,
+}
+
+fn shortest_path_for_centrality<G>(
+    graph: G,
+    node_s: &G::NodeId,
+) -> ShortestPathData
+where
+    G: NodeIndexable
+        + IntoNodeIdentifiers
+        + IntoNeighborsDirected
+        + GraphBase<NodeId = NodeIndex>, // for get() and get_mut()
+{
+    let mut verts_sorted_by_distance: Vec<NodeIndex> = Vec::new(); // a stack
+    #[allow(non_snake_case)]
+    let mut Q: Vec<NodeIndex> = Vec::new(); // a queue
+    let mut predecessors = HashMap::<G::NodeId, Vec<G::NodeId>>::new();
+    let mut sigma = HashMap::<usize, i64>::new();
+    let mut distance = HashMap::<G::NodeId, i64>::new();
```

Comment:
Since we will add `node_count` elements in this dictionaries we can initially create them with this capacity so we don't have to reallocate. (note that we will need type `G` to implement 
 `NodeCount` trait).

```suggestion
    let c = graph.node_count();
    let mut predecessors =
        HashMap::<G::NodeId, Vec<G::NodeId>>::with_capacity(c);
    let mut sigma = HashMap::<usize, i64>::with_capacity(c);
    let mut distance = HashMap::<G::NodeId, i64>::with_capacity(c);
```

I am thinking also if it's better in terms of performance to use here a `Vec` of size `node_bound` (kind of what you did with `betweeness` in `betweenness_centrality`).

Reply:
I find preallocating often doesn't really help. But, this optimization doesn't obfuscate the code, in fact, it signals some intent. A good idea.